### PR TITLE
kubeadm: Make the kubeadm config kinds mutually exclusive

### PR DESCRIPTION
**Pull request by luxas**
**What this PR does / why we need it**:
Right now it would be possible to feed kubeadm with a YAML document with all the four different Config kinds kubeadm at HEAD supports, `MasterConfiguration`, `InitConfiguration`, `JoinConfiguration` and `NodeConfiguration`. This PR makes them mutually exclusive so that kubeadm can know how to handle the config file properly.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
ref: kubernetes/kubeadm#911
Depends on:
 - [x] https://github.com/kubernetes/kubernetes/pull/65776
 - [x] https://github.com/kubernetes/kubernetes/pull/65628
 - [x] https://github.com/kubernetes/kubernetes/pull/65629
 - [x] https://github.com/kubernetes/kubernetes/pull/65631
 - [x] https://github.com/kubernetes/kubernetes/pull/65940
 - [x] https://github.com/kubernetes/kubernetes/pull/65787
 - [x] https://github.com/kubernetes/kubernetes/pull/65945
 - [x] https://github.com/kubernetes/kubernetes/pull/65951

**Special notes for your reviewer**:
Please only review the last commit

**Release note**:

```release-note
NONE
```
@kubernetes/sig-cluster-lifecycle-pr-reviews

### CODE REVIEW

**Comment by luxas** [comment]
The bot seem to not have noticed the rebase, removing the label manually

**Comment by luxas** [comment]
@timothysc please LGTM this when you're ready

File: cmd/kubeadm/app/util/config/common.go [hunk 1]
```diff
@@ -82,10 +82,25 @@ func DetectUnsupportedVersion(b []byte) error {
...
 	// If we find an old API version in this gvk list, error out and tell the user why this doesn't work
+	knownKinds := map[string]bool{}
 	for _, gvk := range gvks {
 		if useKubeadmVersion := oldKnownAPIVersions[gvk.GroupVersion().String()]; len(useKubeadmVersion) != 0 {
 			return fmt.Errorf("your configuration file uses an old API spec: %q. Please use kubeadm %s instead and run 'kubeadm config migrate --old-config old.yaml --new-config new.yaml', which will write the new, similar spec using a newer API version.", gvk.GroupVersion().String(), useKubeadmVersion)
 		}
+		knownKinds[gvk.Kind] = true
+	}
+	// InitConfiguration, MasterConfiguration and NodeConfiguration are mutually exclusive, error if more than one are specified
+	mutuallyExclusive := []string{constants.InitConfigurationKind, constants.MasterConfigurationKind, constants.JoinConfigurationKind, constants.NodeConfigurationKind}
```
Comment (timothysc) [inline cmd/kubeadm/app/util/config/common.go]: Can't we make this more restrictive based on version?  e.g. for 1.12 they can only call is join, etc. 

/approve

Comment (timothysc) [reply cmd/kubeadm/app/util/config/common.go]: I don't feel very strong about this so self-apply lgtm is ok imo.

**Review by timothysc — APPROVED** [review]

**Comment by luxas** [comment]
Adding the LGTM based on https://github.com/kubernetes/kubernetes/pull/65952#discussion_r202747834

> self-apply lgtm is ok imo.

**Pull request merged**